cruise-control-metrics-reporter/src/main/java/com/linkedin/kafka/cruisecontrol/metricsreporter/CruiseControlMetricsReporter.java

```diff
@@ -130,52 +145,98 @@ public void configure(Map<String, ?> configs) {
 
     if (reporterConfig.getBoolean(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_AUTO_CREATE_CONFIG)) {
       try {
-        _newTopic = createNewTopicFromReporterConfig(reporterConfig);
+        _metricsTopic = createMetricsTopicFromReporterConfig(reporterConfig);
         Properties adminClientConfigs = CruiseControlMetricsUtils.addSslConfigs(producerProps, reporterConfig);
         _adminClient = CruiseControlMetricsUtils.createAdminClient(adminClientConfigs);
       } catch (CruiseControlMetricsReporterException e) {
         LOG.warn("Cruise Control metrics topic auto creation was disabled", e);
       }
     }
-
   }
 
-  private NewTopic createNewTopicFromReporterConfig(CruiseControlMetricsReporterConfig reporterConfig)
-  throws CruiseControlMetricsReporterException {
+  protected NewTopic createMetricsTopicFromReporterConfig(CruiseControlMetricsReporterConfig reporterConfig)
+      throws CruiseControlMetricsReporterException {
     String cruiseControlMetricsTopic =
         reporterConfig.getString(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_CONFIG);
     Integer cruiseControlMetricsTopicNumPartition =
         reporterConfig.getInt(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_NUM_PARTITIONS_CONFIG);
     Short cruiseControlMetricsTopicReplicaFactor =
         reporterConfig.getShort(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_REPLICATION_FACTOR_CONFIG);
-
     if (cruiseControlMetricsTopicReplicaFactor <= 0 || cruiseControlMetricsTopicNumPartition <= 0) {
       throw new CruiseControlMetricsReporterException("The topic configuration must explicitly set the replication factor and the num partitions");
     }
+    NewTopic newTopic = new NewTopic(cruiseControlMetricsTopic, cruiseControlMetricsTopicNumPartition, cruiseControlMetricsTopicReplicaFactor);
+
+    Map<String, String> config = new HashMap<>(2);
+    config.put(LogConfig.RetentionMsProp(),
+               Long.toString(reporterConfig.getLong(CruiseControlMetricsReporterConfig.CRUISE_CONTROL_METRICS_TOPIC_RETENTION_MS_CONFIG)));
+    config.put(LogConfig.CleanupPolicyProp(), CRUISE_CONTROL_METRICS_TOPIC_CLEAN_UP_POLICY);
+    newTopic.configs(config);
+    return newTopic;
+  }
+
+  protected void createCruiseControlMetricsTopic() throws ExecutionException {
+    try {
+      CreateTopicsResult createTopicsResult = _adminClient.createTopics(Collections.singletonList(_metricsTopic));
+      createTopicsResult.values().get(_metricsTopic.name()).get(CLIENT_REQUEST_TIMEOUT_MS, TimeUnit.MILLISECONDS);
+      LOG.info("Cruise Control metrics topic {} is created.", _cruiseControlMetricsTopic);
+    } catch (InterruptedException | TimeoutException e) {
+      LOG.warn("Unable to create Cruise Control metrics topic {}.", _cruiseControlMetricsTopic, e);
```

Comment:
Do we need to verify that the exception is a `TopicExistsException` and not due to another reason?

Reply:
In the test I found that `TopicExistsException` will not be directly thrown out here, it will be wrapped by an `ExecutionException`. That's why we do not explicitly handle `TopicExistsException`.

Reply:
Isn't the `ExecutionException` thrown by the call to `get` for the future?

`TopicExistsException` seems to be stored in `CreateTopicsResponse`  (i.e. please see `CreateTopicsRequest#getErrorResponse`), which is handled in `KafkaAdminClient#createTopics` -> `handleResponse`. This completes the future exceptionally.